### PR TITLE
Fix annotations and spans leaking between threads

### PR DESCRIPTION
**Pull request by j-bennet**
Switch global annotations to use `ContextVar` instead of setting values in dask config.

Dependent PR in distributed: https://github.com/dask/distributed/pull/7935

- [x] Closes https://github.com/dask/dask/issues/10340
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

cc @fjetter @crusaderky

### CODE REVIEW

File: dask/base.py [hunk 1]
```diff
@@ -98,6 +102,48 @@ def _restore_excepthook():
 original_excepthook = sys.excepthook
 sys.excepthook = _clean_traceback_hook(sys.excepthook)
 
+_annotations: ContextVar[dict] = ContextVar("annotations", default={})
+
+
+def get_annotations(default_value=None) -> dict | None:
```
Comment (j-bennet) [inline dask/base.py:108]: Initially, I was going to create a module for this code called `dask.annotations`. Unfortunately, we already have `dask.annotations` here:

https://github.com/dask/dask/blob/a6ee367ab76f4b12ab2dcea49e9b91dff9ba63d0/dask/__init__.py#L1

The name would be clashing, so for now, I left it in `base.py`. If you have better suggestions, let me know.

**Comment by j-bennet** [comment]
Initially, I was going to create a module for this code called `dask.annotations`. Unfortunately, we already have `dask.annotations` here:

https://github.com/dask/dask/blob/a6ee367ab76f4b12ab2dcea49e9b91dff9ba63d0/dask/__init__.py#L1

The name would be clashing, so for now, I left it in `base.py`. If you have better suggestions, let me know.

File: dask/base.py [hunk 2]
```diff
@@ -98,6 +102,48 @@ def _restore_excepthook():
...
+
+    Returns
+    -------
+    result : dict | None
+        Dict of annotations, if any
+    """
+    return _annotations.get() or default_value
+
+
+def get_annotation(key: str, *, default_value: Any = None) -> Any:
```
Comment (fjetter) [inline dask/base.py:124]: Is this second function truly necessary? If so, we should choose different function names. The `s` is very easy to miss

**Review by fjetter — COMMENTED** [review]
Can you please add a test for this as well somewhere? https://github.com/dask/dask/issues/10340 provides an example that triggers this leakage w/out spans and I'd like to test this behavior.

File: dask/base.py (hunk 2 above)
Comment (j-bennet) [reply dask/base.py:124]: Not necessary, but it makes for prettier code in `distributed` here:

https://github.com/dask/distributed/blob/1a489575abe7bfbe4459212e44f47bf9dc962487/distributed/spans.py#L77

I can remove it though.

**Review by crusaderky — CHANGES_REQUESTED** [review]
Missing unit tests:
- the code inside `with dask.annotate()` raises
- annotations in two different asyncio tasks or threads[1] running in parallel do not interfere with each other

[1] only need one. Pick whatever is easier/cleaner to write.

**Comment by crusaderky** [comment]
Please add line to `docs/source/api.rst`

File: dask/tests/test_highgraph.py [hunk 3]
```diff
@@ -186,10 +186,19 @@ def test_annotation_and_config_collision():
...
                 assert dask.config.get("foo") == 1
-                assert dask.config.get("annotations") == {"foo": 2, "bar": 3}
+                assert dask.get_annotations() == {"foo": 2, "bar": 3}
+
+
+def test_annotation_cleared_on_error():
+    with dask.annotate(banana=5):
+        with dask.annotate(apple=3):
+            with pytest.raises(ZeroDivisionError):
+                _ = 1 / 0
```
Comment (crusaderky) [inline dask/tests/test_highgraph.py:199]: The exception is being swallowed by pytest.raises and never reaches dask.annotate.

File: dask/tests/test_highgraph.py [hunk 4]
```diff
@@ -186,10 +186,19 @@ def test_annotation_and_config_collision():
     with dask.config.set({"foo": 1}):
         with dask.annotate(foo=2):
             assert dask.config.get("foo") == 1
-            assert dask.config.get("annotations") == {"foo": 2}
+            assert dask.get_annotations() == {"foo": 2}
```
Comment (crusaderky) [inline dask/tests/test_highgraph.py:189]: This test no longer makes sense

**Review by crusaderky — CHANGES_REQUESTED** [review]
Missing unit test:
annotations in two different asyncio tasks or threads[1] running in parallel do not interfere with each other.
[1] only need one. Pick whatever is easier/cleaner to write.

File: docs/source/api.rst [hunk 5]
```diff
@@ -45,7 +45,9 @@ This more advanced API is available in the `Dask distributed documentation
 <https://distributed.dask.org/en/latest/api.html>`_
 
 .. autofunction:: annotate
+.. autofunction:: get_annotations
 .. autofunction:: compute
+.. autofunction:: get_annotations
```
Comment (crusaderky) [inline docs/source/api.rst:50]:
```suggestion
```
I added it already 2 lines above

**Comment by j-bennet** [comment]
> Missing unit test: annotations in two different asyncio tasks or threads[1] running in parallel do not interfere with each other. [1] only need one. Pick whatever is easier/cleaner to write.

I added a test called `test_annotations_threaded`, let me know if that looks right.

File: docs/source/api.rst (hunk 5 above)
Comment (j-bennet) [reply docs/source/api.rst]: Thank you. If you don't mind, let me know next time you're working on my PR at the same time as I do, it was unexpected.

**Comment by j-bennet** [comment]
`test_datetime_std_with_larger_dataset` is failing, but it's unrelated.

**Review by crusaderky — APPROVED** [review]
This is ready to be merged as soon as you revert the temporary CI changes in https://github.com/dask/distributed/pull/7935

**Pull request merged**